In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("/content/CarPrice_Assignment.csv")

In [3]:
df.head()

,car_ID,symboling,CarName,fueltype,aspiration,doornumber,carbody,drivewheel,enginelocation,wheelbase,...,enginesize,fuelsystem,boreratio,stroke,compressionratio,horsepower,peakrpm,citympg,highwaympg,price
0,1,3,alfa-romero giulia,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,13495.0
1,2,3,alfa-romero stelvio,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,16500.0
2,3,1,alfa-romero Quadrifoglio,gas,std,two,hatchback,rwd,front,94.5,...,152,mpfi,2.68,3.47,9.0,154,5000,19,26,16500.0
3,4,2,audi 100 ls,gas,std,four,sedan,fwd,front,99.8,...,109,mpfi,3.19,3.40,10.0,102,5500,24,30,13950.0
4,5,2,audi 100ls,gas,std,four,sedan,4wd,front,99.4,...,136,mpfi,3.19,3.40,8.0,115,5500,18,22,17450.0


In [4]:
df.shape

(205, 26)

In [5]:
df.columns

Index(['car_ID', 'symboling', 'CarName', 'fueltype', 'aspiration',
       'doornumber', 'carbody', 'drivewheel', 'enginelocation', 'wheelbase',
       'carlength', 'carwidth', 'carheight', 'curbweight', 'enginetype',
       'cylindernumber', 'enginesize', 'fuelsystem', 'boreratio', 'stroke',
       'compressionratio', 'horsepower', 'peakrpm', 'citympg', 'highwaympg',
       'price'],
      dtype='object')

In [8]:
df.isnull().sum()

,0
car_ID,0
symboling,0
CarName,0
fueltype,0
aspiration,0
doornumber,0
carbody,0
drivewheel,0
enginelocation,0
wheelbase,0


In [10]:
df.duplicated().sum()

np.int64(0)

In [11]:
#remove unnecessary columns
df = df.drop(columns=["car_ID", "CarName"])

In [12]:
#seperating features and target into separate datasets
x=df.drop(columns=['price'])
y=df['price']

In [13]:
print("Feature data shape:",x.shape)
print("Target data shape:", y.shape)

Feature data shape: (205, 23)
Target data shape: (205,)


In [15]:
print("Numerical columns:")
print(x.select_dtypes(include=["int64", "float64"]).columns)

print("\nCategorical columns:")
print(x.select_dtypes(include=["object"]).columns)

Numerical columns:
Index(['symboling', 'wheelbase', 'carlength', 'carwidth', 'carheight',
       'curbweight', 'enginesize', 'boreratio', 'stroke', 'compressionratio',
       'horsepower', 'peakrpm', 'citympg', 'highwaympg'],
      dtype='object')

Categorical columns:
Index(['fueltype', 'aspiration', 'doornumber', 'carbody', 'drivewheel',
       'enginelocation', 'enginetype', 'cylindernumber', 'fuelsystem'],
      dtype='object')


In [18]:
#convert categorical features into numerical ones using one-hot encoding
x_encoded = pd.get_dummies(x, drop_first=True)

In [19]:
print("Original shape:", x.shape)
print("Encoded shape:", x_encoded.shape)

Original shape: (205, 23)
Encoded shape: (205, 43)


In [21]:
x_encoded.head()

,symboling,wheelbase,carlength,carwidth,carheight,curbweight,enginesize,boreratio,stroke,compressionratio,...,cylindernumber_three,cylindernumber_twelve,cylindernumber_two,fuelsystem_2bbl,fuelsystem_4bbl,fuelsystem_idi,fuelsystem_mfi,fuelsystem_mpfi,fuelsystem_spdi,fuelsystem_spfi
0,3,88.6,168.8,64.1,48.8,2548,130,3.47,2.68,9.0,...,False,False,False,False,False,False,False,True,False,False
1,3,88.6,168.8,64.1,48.8,2548,130,3.47,2.68,9.0,...,False,False,False,False,False,False,False,True,False,False
2,1,94.5,171.2,65.5,52.4,2823,152,2.68,3.47,9.0,...,False,False,False,False,False,False,False,True,False,False
3,2,99.8,176.6,66.2,54.3,2337,109,3.19,3.40,10.0,...,False,False,False,False,False,False,False,True,False,False
4,2,99.4,176.6,66.4,54.3,2824,136,3.19,3.40,8.0,...,False,False,False,False,False,False,False,True,False,False


SPLIT THE DATA INTO TRAIN AND TEST

In [22]:
from sklearn.model_selection import train_test_split

In [23]:
x_train, x_test, y_train, y_test = train_test_split(x_encoded,y,test_size=0.20,random_state=42)

In [24]:
print("x_train shape:", x_train.shape)
print("x_test shape:", x_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

x_train shape: (164, 43)
x_test shape: (41, 43)
y_train shape: (164,)
y_test shape: (41,)


In [29]:
#feature scaling
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

x_train_scaled = scaler.fit_transform(x_train)
x_test_scaled = scaler.transform(x_test)

In [30]:
from sklearn.linear_model import LinearRegression, Ridge, Lasso

In [31]:
linear_model = LinearRegression()

ridge_model = Ridge(alpha=1.0)

lasso_model = Lasso(alpha=1.0, max_iter=10000)

In [32]:
#train the model
linear_model.fit(x_train_scaled, y_train)

ridge_model.fit(x_train_scaled, y_train)

lasso_model.fit(x_train_scaled, y_train)

Lasso(max_iter=10000)

In [33]:
print("Linear Regression model trained")
print("Ridge Regression model trained")
print("Lasso Regression model trained")

Linear Regression model trained
Ridge Regression model trained
Lasso Regression model trained


In [34]:
linear_pred = linear_model.predict(x_test_scaled)
ridge_pred = ridge_model.predict(x_test_scaled)
lasso_pred = lasso_model.predict(x_test_scaled)

In [35]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [36]:
def evaluate_model(model_name, actual, predicted):

    mae = mean_absolute_error(actual, predicted)
    mse = mean_squared_error(actual, predicted)
    rmse = np.sqrt(mse)
    r2 = r2_score(actual, predicted)

    return {"Model": model_name,
        "MAE": mae,
        "MSE": mse,
        "RMSE": rmse,
        "R2": r2 }

In [37]:
linear_results = evaluate_model("Linear Regression", y_test, linear_pred)

ridge_results = evaluate_model( "Ridge Regression", y_test, ridge_pred)

lasso_results = evaluate_model("Lasso Regression", y_test, lasso_pred)

In [38]:
results_df = pd.DataFrame([
    linear_results,
    ridge_results,
    lasso_results
])

results_df

,Model,MAE,MSE,RMSE,R2
0,Linear Regression,2089.382729,8.482008e+06,2912.388793,0.892557
1,Ridge Regression,2104.378812,8.998555e+06,2999.759142,0.886013
2,Lasso Regression,2090.172654,8.545795e+06,2923.319138,0.891749


LINEAR REGRESSION PERFORMED BEST. Its because it had the lowest error values and the highest R² score of about 0.89. Lasso gave very similar results, while Ridge performed slightly worse. Ridge reduced the coefficients, and Lasso could reduce some coefficients to zero. For this dataset, regularization did not improve the model, so Linear Regression generalized best.